Implement GPURenderPassEncoder

kunalmohan opened the pull request:
Add webidls for GPURenderPassEncoder and  GPURenderEncoderBase and
implement relevant methods.

<!-- Please describe your changes on the following line: -->
r?@kvark

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

kunalmohan committed:
Implement GPURenderPassEncoder
Add webidls for GPURenderPassEncoder and  GPURenderEncoderBase and
implement relevant methods.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webidls/GPURenderEncoderBase.webidl, components/script/dom/gpurenderpipeline.rs, components/script/dom/webidls/GPURenderPassEncoder.webidl, components/script/dom/gpucomputepassencoder.rs, components/script/dom/gpudevice.rs and 5 more
 * @kichjang: components/script/dom/webidls/GPURenderEncoderBase.webidl, components/script/dom/gpurenderpipeline.rs, components/script/dom/webidls/GPURenderPassEncoder.webidl, components/script/dom/gpucomputepassencoder.rs, components/script/dom/gpudevice.rs and 5 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

kunalmohan commented:
```
error[E0432]: unresolved import `crate::dom::bindings::codegen::UnionTypes::GPULoadOpOrRangeEnforcedUnsignedLong`
  --> components/script/dom/gpucommandencoder.rs:13:5
   |
13 |     GPULoadOpOrRangeEnforcedUnsignedLong as GPULongLoad,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `GPULoadOpOrRangeEnforcedUnsignedLong` in `dom::bindings::codegen::UnionTypes`
```
This did not show up during my local build :(

The thing is I have this dictionary in `GPUCommandEncoder.webidl`
```
dictionary GPURenderPassDepthStencilAttachmentDescriptor {
    required GPUTextureView attachment;

    required (GPULoadOp or float) depthLoadValue;
    required GPUStoreOp depthStoreOp;
    boolean depthReadOnly = false;

    required (GPULoadOp or GPUStencilValue) stencilLoadValue;
    required GPUStoreOp stencilStoreOp;
    boolean stencilReadOnly = false;
};
```
I need to match on `stencilLoadValue` and `depthLoadValue` for which I need to import their types. Is there someplace else I should import them?

jdm commented:
This is really strange.
@bors-servo try=linux

bors-servo commented:
:hourglass: Trying commit 4b6316c6be3a498c0c74c70f87af3092c1810d2d with merge 387495a5174c076a656116a3ffcd4d9c9999b3d0...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/NjhfP-fQR1ekUL1nMm3WrQ)

jdm commented:
It makes no sense to me that your local build does not reproduce this build error. My only advice at this point is to do the following:
* make a build error occur in one of gpuadapter.rs by removing an argument from one of the trait methods
* using the path to the generated file that is present in the error message you receive, find the generated UnionTypes.rs that is in the generated bindings' parent directory
* gist that file so we can better understand if GPULoadOpOrRangeEnforcedUnsignedLong is actually present or not

kunalmohan commented:
@jdm https://gist.github.com/kunalmohan/59b4c4dc2beb3fb789da2ed2507a9bcf

kvark commented on components/script/dom/gpurenderpassencoder.rs:
```diff
@@ -0,0 +1,283 @@
+        raw_pass: RawPass,
+        parent: &GPUCommandEncoder,
+        device: WebGPUDevice,
+    ) -> GPURenderPassEncoder {
```
nit: `-> Self`, can be used in other places too

kvark reviewed (approved):
Looks great! I have a few questions, no blocking concerns.

jdm commented:
Super weird. I built your branch locally and don't see the build error. What OS are you using?

jdm commented:
@bors-servo try=mac retry

bors-servo commented:
:hourglass: Trying commit 4b6316c6be3a498c0c74c70f87af3092c1810d2d with merge 4a17df686a9e5825fa934644f12b4a0cbec98274...

kunalmohan commented on components/script/dom/gpucomputepassencoder.rs:
```diff
@@ -21,28 +22,33 @@ use webgpu::{
     label: DomRefCell<Option<DOMString>>,
     #[ignore_malloc_size_of = "defined in wgpu-core"]
-    raw_pass: RefCell<Option<RawPass>>,
+    raw_pass: DomRefCell<Option<RawPass>>,
```
It's isn't necessary. RefCell would work too. Its just that DomRefCell has a gives useful output for dynamic borrow failures behind the feature "refcell-backtrace".

kunalmohan commented on components/script/dom/gpucomputepassencoder.rs:
```diff
@@ -21,28 +22,33 @@ use webgpu::{
     reflector_: Reflector,
     #[ignore_malloc_size_of = "defined in webgpu"]
     channel: WebGPU,
+    device: WebGPUDevice,
```
Yep. I'll remove this.

kunalmohan commented on components/script/dom/gpurenderpassencoder.rs:
```diff
@@ -0,0 +1,283 @@
+    /// https://gpuweb.github.io/gpuweb/#dom-gpurenderencoderbase-setindexbuffer
+    fn SetIndexBuffer(&self, buffer: &GPUBuffer, offset: u64, size: u64) {
+        let s;
+        if size == 0 {
```
I haven't updated wgpu-core recently, so no. I'll do so now.

kunalmohan commented:
I am on Ubuntu 18.04.

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/WmcFoCisQV2UX6nc7BSj_Q)
State: approved= try=True

jdm commented:
As a possible workaround, does anything change if you make a typedef for that union and import the typedef name in rust instead?

kunalmohan commented:
I have modified that particular type to use typedef, as suggested, to check if the build succeeds.
I haven't yet addressed all the comments from the review yet- wgpu-core is yet to be updated- so please don't merge.

kunalmohan commented:
So build succeeds now \o/. I don't know why that particular type was throwing the error while the other two didn't (Maybe it was cursed :P).
I'll proceed with this if that's not a problem, and push the other changes too.

kunalmohan committed:
Update wgpu-core and wgpu-types

kunalmohan commented:
@kvark PTAL!

jdm commented:
@bors-servo r=kvark

bors-servo commented:
:pushpin: Commit 3b5ede1 has been approved by `kvark`

<!-- @bors-servo r=kvark 3b5ede153d61c2c033326b5df2417e3757dba17b -->

bors-servo commented:
:hourglass: Testing commit 3b5ede153d61c2c033326b5df2417e3757dba17b with merge 643c614d6895601d4b13f82c54f9ce273c052362...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ZgW-nSdHSDmNzj6Na-x88g)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 3b5ede153d61c2c033326b5df2417e3757dba17b with merge 0281acea955085aec01a4fc6da5f89f326a14842...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/KI9JAm2qSP6rOWmk7KGbWg)
Approved by: kvark
Pushing 0281acea955085aec01a4fc6da5f89f326a14842 to master...
